examples/llm/glem.py

```diff
@@ -0,0 +1,440 @@
+          f'Test: {test_acc:.4f}')
+
+    # EM iterations
+
+    em_phase = em_order
+    """
+    We run E-step(LM training) and M-Step(GNN training) alternatively in each
+    em iterations, so the total number of iterations is num_em_iter * 2 and
+    we switch the em_phase at end of each iteration in following loop
+    """
+    for em_it in range(1, num_em_iters * 2 + 1):
+        pseudo_labels = preds.argmax(dim=-1)
+        best_val_acc = final_test_acc = 0.0
+        print(f'EM iteration: {em_it}, EM phase: {em_phase}')
+        optimizer = load_model(em_phase)
+        num_epochs = lm_epochs
+        train_loader = text_train_loader
+        test_loader = text_data_loader
+        early_stopping = 0
+        if em_phase == 'gnn':
+            train_loader = graph_train_loader
+            num_epochs = gnn_epochs
+            test_loader = subgraph_loader
+        for epoch in range(1, num_epochs + 1):
+            acc, loss = model.train(em_phase, train_loader, optimizer,
+                                    pseudo_labels, epoch, True, verbose)
+            if epoch >= 5 or epoch == num_epochs:
+                cur_preds = model.inference(em_phase, test_loader,
+                                            verbose=verbose)
+                train_acc, val_acc, test_acc = evaluate(
+                    cur_preds, ['train', 'valid', 'test'])
+
+                print(f'Train: {train_acc:.4f}, Val: {val_acc:.4f}, '
+                      f'Test: {test_acc:.4f}')
+
+                if val_acc <= best_val_acc:
+                    early_stopping += 1
+                    if early_stopping > patience:
+                        print(f'''Early stopped by Epoch: {epoch}, \
+                            Best acc: {final_test_acc}''')
+                        break
+                else:
+                    best_val_acc = val_acc
+                    final_test_acc = test_acc
+                    preds = cur_preds
+
+        if em_phase == 'gnn':
+            gnn_test_acc = max(gnn_test_acc, final_test_acc)
+            model.gnn = model.gnn.to('cpu', non_blocking=True)
+            em_phase = 'lm'
+        else:
+            lm_test_acc = max(lm_test_acc, final_test_acc)
+            model.lm = model.lm.to('cpu', non_blocking=True)
+            em_phase = 'gnn'
+        torch.cuda.empty_cache()
+    print(f'Best GNN acc: {gnn_test_acc}, LM acc: {lm_test_acc}')
+    print('============================')
+    end_time = time.time()
+    running_time = (end_time - start_time) / 3600
+    print(f'Total running time: {running_time:.2f} hours')
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='GLEM Example:')
+    parser.add_argument('--gpu', type=int, default=0)
+    parser.add_argument('--num_runs', type=int, default=10,
+                        help='number of runs')
+    parser.add_argument('--num_em_iters', type=int, default=1,
+                        help='number of iterations')
```

Comment:
Is `1` a good default value for EM iterations?

Reply:
Yes, I have tested it till 5, mostly 1-2 EM iteration can perform very well on ogbn-products and ogbn-arxiv